[AutoDiff] Handle materializing adjoints with non-differentiable fields

jkshtj opened the pull request:
This PR fixes a compiler crasher in AutoDiff. 

The compiler used to crash while generating a pullback for differentiable functions that take a differentiable input, whose tangent vector contains non-differentiable fields.

Changes in this PR fix the issue by specializing adjoint materialization logic for user-defined tangent vectors containing non-differentiable fields.

Fixes #66522

jkshtj commented:
@asl @BradLarson could one of you invoke swift-ci tests?

jkshtj commented on test/AutoDiff/compiler_crashers_fixed/issue-66522-pullback-generation-when-tangentvector-of-input-contains-nondifferentiable-fields.swift:
```diff
@@ -0,0 +1,67 @@
+// NOW WHAT?
+// Is this edge case something that we need to handle?
+// 
+// @differentiable(reverse)
```
@rxwei Could you please take a look at the comments on this commented-out differentiable function? 

I may have discovered an edge-case and I wanted to confirm whether or not it's something we need to handle and/or fix.

asl commented:
I am confused. I think the discussion was that we need to fix adjoint generation for `struct_extract` instruction. Are there other usecases besides `struct_extract` that should be covered?

What is the test coverage for newly added code?

asl commented:
@swift-ci please test

jkshtj commented:
> I am confused. I thought the discussion was that we need to fix adjoint generation for `struct_extract` instruction.

I think I may have misunderstood how to go about fixing the issue. Are you suggesting that we never add non-differentiable fields to the adjoint for `struct_extract`?

jkshtj commented:
> Are there other usecases besides `struct_extract` that should be covered?

If we do make the fix in the adjoint generation code, then I believe we'd have to at least do the same for `tuple_extract`. There may be other instructions as well, but I'd need to take a closer look.

jkshtj commented:
> What is the test coverage for newly added code?

In unit tests none. 

In the end to end tests, I added a tests for the path where we need to materialize an adjoint with non-differentiable fields. I should perhaps also have added a test for materializing normal adjoints with all differentiable fields.

asl commented:
> I think I may have misunderstood how to go about fixing the issue. Are you suggesting that we never add non-differentiable fields to the adjoint for `struct_extract`?

There are multiple questions that need to be answered:

1. Are there any other instructions that are affected here. Besides `struct_extract`? For example, you've added handling of adjoint buffers in addition to values. Are there any testcases for this code?

2. Next, the current adjoint generation for `struct_extract` is essentially:
```
  ///   Original: y = struct_extract x, #field
  ///    Adjoint: adj[x] += struct (0, ..., #field': adj[y], ..., 0)
  ///                                       ^~~~~~~
  ///                     field in tangent space corresponding to #field
```

So, instead of materalizing that bunch of zeros, why we simply cannot do the following:
```
adj[x].field += adj[y]
```

After all, we know that things should conform to `AdditiveArithmetic` and therefore `a += 0` should be equal to `a`.

asl commented:
> In unit tests, none.

Please ensure that all codepaths are covered by tests.

jkshtj commented:
>     1. Are there any other instructions that are affected here. Besides `struct_extract`? For example, you've added handling of adjoint buffers in addition to values. Are there any testcases for this code?

Since the adjoint materialization code is shared by all instructions, we shouldn't need to handle things separately for different instructions. As for the handling of adjoint buffer handling, it looks like that code isn't actually called anywhere.

jkshtj commented:
>     2. Next, the current adjoint generation for `struct_extract` is essentially:
> 
> 
> ```
>   ///   Original: y = struct_extract x, #field
>   ///    Adjoint: adj[x] += struct (0, ..., #field': adj[y], ..., 0)
>   ///                                       ^~~~~~~
>   ///                     field in tangent space corresponding to #field
> ```
> 
> So, instead of materalizing that bunch of zeros, why we simply cannot do the following:
> 
> ```
> adj[x].field += adj[y]
> ```
> 
> After all, we know that things should conform to `AdditiveArithmetic` and therefore `a += 0` should be equal to `a`.

Ah, indeed this is what I tried to do first and I ran into an issue with `TangentBuilder`'s `emitInPlaceAdd` method, which takes the lhs as a buffer. This was causing SIL verification failures for '+='-ing trivial types such as `Double`, with the following error message - 
```
SIL verification failed: operand of 'apply' doesn't match function input type
  $Double
  $*Double
Verifying instruction:
   %0 = argument of bb0 : $Double                 // user: %8
     %5 = struct_element_addr %1 : $*P<Double>, #P.value // user: %8
     %6 = witness_method $Double, #AdditiveArithmetic."+=" : <Self where Self : AdditiveArithmetic> (Self.Type) -> (inout Self, Self) -> () : $@convention(witness_method: AdditiveArithmetic) <τ_0_0 where τ_0_0 : AdditiveArithmetic> (@inout τ_0_0, @in_guaranteed τ_0_0, @thick τ_0_0.Type) -> () // user: %8
     %7 = metatype $@thick Double.Type            // user: %8
->   %8 = apply %6<Double>(%5, %0, %7) : $@convention(witness_method: AdditiveArithmetic) <τ_0_0 where τ_0_0 : AdditiveArithmetic> (@inout τ_0_0, @in_guaranteed τ_0_0, @thick τ_0_0.Type) -> ()
```

It looks like I may have misinterpreted the meaning of the error, however. And I think I simply needed to change the type of the input arguments. 

I can definitely try this out.

jkshtj commented:
> > In unit tests, none.
> 
> Please ensure that all codepaths are covered by tests.

Would you know of an existing unit test that tests auto diff code? I wasn't quite sure how I can setup `PullbackCloner` for unit testing.

asl commented:
> > ```
> > 1. Are there any other instructions that are affected here. Besides `struct_extract`? For example, you've added handling of adjoint buffers in addition to values. Are there any testcases for this code?
> > ```
> 
> Since the adjoint materialization code is shared by all instructions, we shouldn't need to handle things separately for different instructions. As for the handling of adjoint buffer handling, it looks like that code isn't actually called anywhere.

Allright, but what are those "different instruction"? Besides `tuple_extract`? Looks like you're adding whole bunch of generic code without any tests just to fix one particular corner case. Differential paths are known to have less-than-optimal test coverage and this usually cause all kinds of issues when the code is actually called in some obscure case. So we try to improve the test coverage, not to reduce it.

asl commented:
> Would you know of an existing unit test that tests auto diff code? I wasn't quite sure how I can setup `PullbackCloner` for unit testing.

`test/AutoDiff` contain the tests for different aspects of autodiff code.

asl commented:
> This was causing SIL verification failures for '+='-ing trivial types such as `Double`, with the following error message -

The `lhs` is a struct somewhere. So you should be able to take the address of its field and pass to `+=` method.

jkshtj commented:
> > Would you know of an existing unit test that tests auto diff code? I wasn't quite sure how I can setup `PullbackCloner` for unit testing.
> 
> `test/AutoDiff` contain the tests for different aspects of autodiff code.

Oh, I think I was mistaking unit tests for the ones that occur in the [unittests](https://github.com/apple/swift/tree/main/unittests) directory in the Swift repo.

jkshtj commented:
> > This was causing SIL verification failures for '+='-ing trivial types such as `Double`, with the following error message -
> 
> The `lhs` is a struct somewhere. So you should be able to take the address of its field and pass to `+=` method. However, in your particular case it was rhs that caused the assertion as the argument is passed indirect.

Yeah it was the right one. I was a little bit surprised because I was thinking that only the left one needs to be taken as a pointer (or by reference). But I'm guessing this more generalized SIL definition exists to accomodate for different types that may want to conform to AdditiveArithmetic?

jkshtj commented:
> > > ```
> > > 1. Are there any other instructions that are affected here. Besides `struct_extract`? For example, you've added handling of adjoint buffers in addition to values. Are there any testcases for this code?
> > > ```
> > 
> > 
> > Since the adjoint materialization code is shared by all instructions, we shouldn't need to handle things separately for different instructions. As for the handling of adjoint buffer handling, it looks like that code isn't actually called anywhere.
> 
> Allright, but what are those "different instruction"? Besides `tuple_extract`? Looks like you're adding whole bunch of generic code without any tests just to fix one particular corner case. Differential paths are known to have less-than-optimal test coverage and this usually cause all kinds of issues when the code is actually called in some obscure case. So we try to improve the test coverage, not to reduce it.

I see the point. Indeed, I missed a number of tests that could be added with this change. I was thinking we can start by covering pullbacks for instructions listed [here](https://github.com/apple/swift/blob/main/docs/DifferentiableProgrammingImplementation.md#derivative-code-generation-details). 

Can you think of any others that I should also test?

jkshtj commented:
>     2. Next, the current adjoint generation for `struct_extract` is essentially:
> 
> 
> ```
>   ///   Original: y = struct_extract x, #field
>   ///    Adjoint: adj[x] += struct (0, ..., #field': adj[y], ..., 0)
>   ///                                       ^~~~~~~
>   ///                     field in tangent space corresponding to #field
> ```
> 
> So, instead of materalizing that bunch of zeros, why we simply cannot do the following:
> 
> ```
> adj[x].field += adj[y]
> ```
> 
> After all, we know that things should conform to `AdditiveArithmetic` and therefore `a += 0` should be equal to `a`.

@asl I've incorporated this feedback about adjoint materialization and am creating a new PR revision for some early feedback. 

I'm still working on adding test coverage but am struggling a bit to understand what exactly the tests should be doing. I'll post my questions about tests in another comment shortly.

jkshtj commented:
@rxwei Could you also please take a look at the approach and comment whether it makes sense and whether I am making changes at the right place?

asl commented:
> @asl I've incorporated this feedback about adjoint materialization and am creating a new PR revision for some early feedback.

Have you forgotten to push your new changes?

jkshtj commented:
> > @asl I've incorporated this feedback about adjoint materialization and am creating a new PR revision for some early feedback.
> 
> Have you forgotten to push your new changes?

I don't think so. I amended the existing commit, however, instead of creating a new one.

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -376,38 +481,114 @@ class PullbackCloner::Implementation final
     case AdjointValueKind::Concrete:
       auto concreteVal = val.getConcreteValue();
-      builder.emitStoreValueOperation(loc, concreteVal, destAddress,
+      // `val` needs to be an owned value for storing it into `destAddress`,
```
Do we have testcase for this?

jkshtj replied:
I do, locally. Haven't pushed it yet.

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -362,6 +380,93 @@ class PullbackCloner::Implementation final
+              builder.createAllocStack(loc, adjField.getType())->getResult(0);
+          materializeAdjointIndirect(adjField, rhsAdjEltBuf, loc);
+
+          // lhsAdjEltBuf += rhsAdjEltBuf
```
I think it's up to the caller to decide what to do with adjoint value. Can you fold these two into adjoint generation of `tuple_extract` and `struct_extract` correspondingly?

asl replied:
To be explicit. What I'm suggesting:

For `y = struct_extract x, #field` the adjoint generation would be:
 1. You take adjoint of 'x'
 1. You take adjoint of 'y'
 1. You take address of #field in x'
 1. You do x'.#field += y'

asl commented on test/AutoDiff/SILOptimizer/pullback_generation.swift:
```diff
@@ -0,0 +1,129 @@
+    v.a
+}
+
+// CHECK-LABEL: [ORIG]  %2 = struct_extract %0 : $PiecewiseMaterializable, #PiecewiseMaterializable.a // user: %3
```
Maybe it will make sense to check the generated SIL and not the debug output? After all, you can do `-Xllvm -sil-print-after=differentiation`

jkshtj replied:
Yeah sure.

asl commented on test/AutoDiff/compiler_crashers_fixed/issue-66522-pullback-generation-when-tangentvector-of-input-contains-nondifferentiable-fields.swift:
```diff
@@ -0,0 +1,67 @@
+// RUN: %target-swift-frontend -emit-sil -O %s
```
should we XFAIL it then?

jkshtj replied:
Ah, why? This does not cause a crash anymore.

asl replied:
Does the comment at line 35 apply to the whole testcase? Or just the commented out lines?

jkshtj replied:
Just the commented out lines. Sorry, if that wasn't clear. 

I wanted to get Richard's comments on what seems like an edge case when using custom tangent vectors. But I can perhaps decouple it from this PR and handle it separately.

asl replied:
+1

asl commented on test/AutoDiff/SILOptimizer/pullback_generation.sil:
```diff
@@ -0,0 +1,155 @@
+  return %2 : $X                                  
+}
+
+// CHECK-LABEL: [ORIG]  %1 = struct_extract %0 : $Y, #Y.a               // user: %2
```
I would rather do `CHECK-LABEL` for function name + see comment below

jkshtj replied:
Will change this to use the actual SIL after the differentiation pass instead.

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -1168,22 +1168,35 @@ class PullbackCloner::Implementation final
         break;
       case AdjointValueKind::Concrete:
       case AdjointValueKind::Aggregate: {
```
Right. Please do.

rxwei commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -1162,28 +1162,41 @@ class PullbackCloner::Implementation final
-        addAdjointValue(bb, sei->getOperand(),
-                        makeAggregateAdjointValue(tangentVectorSILTy, eltVals),
-                        loc);
+        llvm_unreachable("Adjoint value kind of the input to `struct_extract` "
```
With this, you are now forcing all `struct_extract` to materialize the adjoint. To me it seems unnecessary and may lead to performance issues for common cases where there is no non-differentiable fields of a `TangentVector`.

Can we use a type-driven approach instead?  It is still fine to create an `Aggregate` adjoint value or use the `struct` instruction, so long as the following conditions are met:
1. `TangentVector` is defined in the same compilation unit of differentiation.
2. All of `TangentVector`’s stored properties have a same-name differentiable property in the original type.

Or better, you could create a new `AdjointValueKind` called `AddElement`.  When the above conditions aren't met, just create a `AddElement(adj[x], #field, adj[y])` adjoint value. This way the whole thing can be kept symbolic and won't be materialized unless it's absolutely necessary.

jkshtj replied:
Thanks for reviewing this Richard! I have a follow up question.

> With this, you are now forcing all struct_extract to materialize the adjoint. To me it seems unnecessary and may lead to performance issues for common cases where there is no non-differentiable fields of a TangentVector.

Aren't all adjoints eventually going to materialize, before the end of the differentiation pass, given the differentiation pass is run once and atomically(right?) with no other optimization passes run in between that may perform some peephole optimizations?

rxwei replied:
> Aren't all adjoints eventually going to materialize

No, the purpose of the `AdjointValue` data structure is to prevent eager materialization and try to fold the data structure as much as possible. This is why we have special handling of different `AdjointValueKind`s for most instructions. There's a lot of cases of creating a struct of all zeros but one element, then using just one element from it in subsequent code.

jkshtj replied:
That makes sense Richard. I also discussed this with Brad, Anton and others and it seems like the right direction, at the moment. Let me make the changes you suggested and get a new revision out.

jkshtj replied:
@rxwei I was wondering something about this new adjoint kind we need to create. Could it just be a closure that simply returns a `SILValue`? I'm wondering if a new adjoint kind like `AddElement(adj[x], #field, adj[y])` is too specific and perhaps we can create something more general that works for all adjoints that may need some custom materialization logic?

jkshtj replied:
Another follow-up question. 

Would you be able to comment on what the representation, in the actual implementation of the new adjoint kind you're suggesting (`AddElement(adj[x], #field, adj[y])`), should look like? The 2 I can think of are:
1. Store the original `struct_extract` or `tuple_extract` instruction.
2. Store the adjoints of `x` and `y` and the `VarDecl` for the field in `y = struct_extract x, #field`

asl replied:
You need to be able to symbolically evaluate the adjoint in case it would be needed for materialization. Note that both `x` and `y` here could also be symbolic adjoints. So, the `2.` looks like a way to go for me.

jkshtj commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -2993,13 +3092,60 @@ AdjointValue PullbackCloner::Implementation::accumulateAdjointsDirect(
       }
       return makeAggregateAdjointValue(lhsVal->getType(), newElements);
     }
+    // x + (zeroBase, index, eltToAdd) => (x.0, ..., x.index + eltToAdd, ...,
```
@rxwei @asl I looked into adding a unit test (I'm referring to actual unit tests testing C++ classes/functions directly) for this function, but ran into some issues due to the internal state changes in PullbackCloner, that happen during this its call. 

Do you guys see any issues if I create a PullbackCloner constructor initializing minimal state, required for this test?

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -400,12 +458,45 @@ class PullbackCloner::Implementation final
+      auto baseAdjoint = addElementValue->baseAdjoint;
+      auto eltToAdd = addElementValue->eltToAdd;
+
+      builder.emitZeroIntoBuffer(loc, destAddress, IsInitialization);
```
Shouldn't you do indirect materialization of `baseAdjoint` into `destAddress?

jkshtj replied:
You're right. Changed it.

asl commented on include/swift/SILOptimizer/Differentiation/AdjointValue.h:
```diff
@@ -38,10 +40,25 @@ enum AdjointValueKind {
   /// A concrete SIL value.
   Concrete,
+
+  /// A special adjoint, made up of 2 adjoints -- a base adjoint and an element
```
Do we really need to provide a "clarification" here? This might confuse the reader. What is confusing is `base adjoint` term. It's not quite so. We're essentially using only type (and index there). Also, do we really need to use `Add` here? I think the semantics is essentially `(0, ..., 0, el', 0, ..., 0)`, so the non-trivial combination of zero and aggregate adjoints.

jkshtj replied:
Side-stepping the naming suggestions for a second, what I understood about the semantics of this new adjoint kind is -
1. It is a combination (by addition) of 2 adjoints, let's call them LHS and RHS for now. 
2. The LHS will always be a zeroed aggregate adjoint.
3. The RHS will be the adjoint for one of the fields of LHS. 

Symbolically, we have -
```
AddElement(LHS, #field_or_index, RHS) == (materialization) ==> AddElement.#field_or_index += RHS
```
There might be questions/comments about how we're accumulating this new adjoint with others, but I think establishing a common understanding of its semantics first might be helpful 🙂.

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -2993,13 +3092,60 @@ AdjointValue PullbackCloner::Implementation::accumulateAdjointsDirect(
       }
       return makeAggregateAdjointValue(lhsVal->getType(), newElements);
     }
+    // x + (zeroBase, index, eltToAdd) => (x.0, ..., x.index + eltToAdd, ...,
```
why can't you simply call `accumulateAdjoints` on `x` and `base`, and then wrap the result with `AddElement` again? This way you'd evaluate everything symbolically and only materialize the real value (emit addition) when necessary. It will also collapse zeros on fly.

jkshtj replied:
Fixed in the latest revision.

jkshtj commented:
@asl or @BradLarson could one of you kick of swift-ci?

asl commented:
@swift-ci please test

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -2994,13 +3093,70 @@ AdjointValue PullbackCloner::Implementation::accumulateAdjointsDirect(
       return makeAggregateAdjointValue(lhsVal->getType(), newElements);
     }
+    // x + (baseAdjoint, index, eltToAdd)
+    // There are 2 possible cases -
```
The whole intention of new adjoint kind is to prevent extraneous materialization of adjoints. Here it seems you're materializing them. Also, why do you need these explicit cases? Why can't you simply call `accumulateAdointsDirect(base, rhs)`, wrap result into `AddElement(, eltToAdd)` and return it? All that zero folding, etc. will be already handled for you.

jkshtj replied:
I see the issue here and I see what you were saying earlier. I was trying to constrain AddElement adjoint's base adjoint for no reason, it seems. It looks like it can be anything, which should truly let us defer materialization until `materializationAdjoint` is called.

asl commented on include/swift/SILOptimizer/Differentiation/AdjointValue.h:
```diff
@@ -38,10 +40,17 @@ enum AdjointValueKind {
+
+  /// A special adjoint, made up of 2 adjoints -- a base adjoint and an element
+  /// adjoint to add to it. This case exists to avoid eager materialization of
+  /// `Aggregate` and `Zero` adjoints upon addition.
```
Can the base be `AddElement` as well?

jkshtj replied:
Updated in the latest revision.

jkshtj commented:
@BradLarson @asl could one of you kick off swift-ci please?

asl commented:
@swift-ci please test

rxwei commented on lib/SILOptimizer/Differentiation/AdjointValue.cpp:
```diff
@@ -0,0 +1,71 @@
+    s << "]";
+    break;
+  }
+}
```
Nit: Add a new line at the end like other files?

asl commented on lib/SILOptimizer/Differentiation/AdjointValue.cpp:
```diff
@@ -0,0 +1,71 @@
+//===--- AdjointValue.cpp - Helper class for differentiation ----*- C++
+//-*---===//
```
Extra newline?

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -400,12 +422,59 @@ class PullbackCloner::Implementation final
+      SmallVector<AddElementValue *, 4> addEltAdjValues;
+
+      do {
+        auto addElementValue = baseAdjoint.getAddElementValue();
```
Maybe add some assertions here? After all, all these adoints must be either tuple-type or struct-like.

asl reviewed (approved):
Thanks! This variant seems to provide the intended functionality. See small comments here and there

jkshtj commented:
@rxwei @asl Thanks a lot for reviewing this PR so far.

I have fixed the nits. Could you please take a look?

asl commented:
> @rxwei @asl Thanks a lot for reviewing this PR so far!
> 
> I have fixed the nits. Could you please take a look?

See new comment.

jkshtj commented:
> See new comment. 
Did you forget to publish the comment 😅?

> Also, will you please squash the commits?
Sure.

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -430,6 +430,8 @@ class PullbackCloner::Implementation final
     }
     case AdjointValueKind::AddElement: {
       auto baseAdjoint = val;
+      assert(baseAdjoint.getType().is<TupleType>() ||
```
This is not what should be checked here. This assertion is enforced during `AddElement` construction. Here you need to ensure that all nested adjoints are of the same kind: either all tuples or all structs, but not the mixture.

jkshtj replied:
I see. Sorry I sent out that last revision in haste. Fixed in the latest revision and squashed commits.

asl commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -400,12 +422,61 @@ class PullbackCloner::Implementation final
+        addEltAdjValues.push_back(addElementValue);
+        baseAdjoint = addElementValue->baseAdjoint;
+        assert(baseAdjointType == baseAdjoint.getType());
+      } while (baseAdjoint.getKind() == AdjointValueKind::AddElement);
```
Do we have tests for this loop, btw? Maybe I'm missing something, but all loops are doing single struct/tuple extract?

jkshtj replied:
Not explicit tests but the other tests I have added do use this code path -- I needed to fix new test failures after making these changes. But I think it's a good idea to add explicit tests for this.

jkshtj replied:
@asl I have added tests exercising the nested add element adjoint materialization path.

asl reviewed (approved):
Thanks!

asl commented:
@swift-ci please test

jkshtj commented:
@asl Had to make some changes after I discovered some small bugs while compiling and running internal test suite. I have added tests for these cases as well.

jkshtj committed:
[AutoDiff] Handle materializing adjoints with non-differentiable fields

Fixes #66522

jkshtj committed:
[AutoDiff] Fixes small bugs in the PR fixing #66522

During internal testing we discovered 2 more bugs -
1. The element adjoint of a struct_extract can itself be an AddElement.
2. Indirect concrete adjoint materialization was missing a copy operation.

This commit fixes these bugs and adds relevant test cases.

asl reviewed (approved):
Makes sense, thanks!

asl commented:
@swift-ci please test

jkshtj commented:
@asl looks like the "Swift Test Linux Platform" failed due to some network issues. Can we restart them?

BradLarson commented:
@swift-ci Please test Linux platform

asl commented:
@swift-ci please test linux

jkshtj commented:
Huh I see the following error due to which the Linux tests failed -

> ERROR: Error fetching remote repo 'origin'

Seems like something may be up with the Linux build hosts?

asl commented:
@shahmishal Is CI down?
